File: python/src/patterns/cqrs/read_model.py
from typing import Dict
from .events import DomainEvent, UserCreated, UserNameUpdated

class ReadModel:
    def __init__(self) -> None:
        self.users: Dict[str, dict] = {}
    def project(self, e: DomainEvent) -> None:
        if isinstance(e, UserCreated):
            self.users[e.id] = {"id": e.id, "name": e.name, "email": e.email}
        elif isinstance(e, UserNameUpdated):
            u = self.users.get(e.id)
            if u: u["name"] = e.name
    def get_user(self, id: str):
        return self.users.get(id)
    def all(self):
        return list(self.users.values())

File: typescript/src/patterns/cqrs/ReadModel.ts
import type { DomainEvent } from './Events.js';

export interface User { id: string; name: string; email: string; }

export class ReadModel {
  private users = new Map<string, User>();

  project(e: DomainEvent) {
    switch (e.type) {
      case 'UserCreated': this.users.set(e.id, { id: e.id, name: e.name, email: e.email }); break;
      case 'UserNameUpdated': {
        const u = this.users.get(e.id);
        if (u) u.name = e.name;
        break;
      }
    }
  }
  getUser(id: string) { return this.users.get(id); }
  all() { return Array.from(this.users.values()); }
}
